java/dataset/src/main/java/org/apache/arrow/dataset/substrait/SubstraitConsumer.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.dataset.substrait;
+
+import org.apache.arrow.c.ArrowArrayStream;
+import org.apache.arrow.c.Data;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.ipc.ArrowReader;
+
+/**
+ * Java binding of the C++ ExecuteSerializedPlan.
+ */
+public class SubstraitConsumer implements AutoCloseable {
+  private final BufferAllocator allocator;
+
+  public SubstraitConsumer(BufferAllocator allocator) {
+    this.allocator = allocator;
+  }
+
+  /**
+   * Execute Substrait Plan to read Schema and ArrowRecordBatches.
+   *
+   * @param substraitPlan the JSON Substrait plan.
+   * @return the ArrowReader to iterate for record batches.
+   */
+  public ArrowReader runQuery(String substraitPlan) {
```

Comment:
Plans should be byte[] or ByteBuffer?

Reply:
It seems this assumes JSON plans which is probably not what people will be working with

Reply:
Please if yu could help me, What will be my options to exchange a Substrait Plan (`Java byte`) to `std::shared_ptr<arrow::Buffer>` called by JNI?

Reply:
JNI functions can read byte arrays, you can then copy into an Arrow buffer (this is probably good anyways to avoid too many cross-boundary dependencies)

Reply:
I just create this `ByteBuffer` with Substrait Plan 
```
ByteBuffer directByteBuffer = ByteBuffer.allocateDirect(64);
directByteBuffer.put("DEMO_SUBSTRAIT_PLAN".getBytes(StandardCharsets.UTF_8)); // protoPlan.toByteArray();
```

JNI Wrapper:
Recover ByteArray with:
```
jbyte *buff = (jbyte *) env->GetDirectBufferAddress(plan);
```

How I could transform that `jbyte` into `std::shared_ptr<arrow::Buffer>`

Reply:
Get the capacity, allocate a mutable buffer, memcpy the result

Reply:
changed